// firebase-tools/lib/serve/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionsServer = void 0;
const path = require("path");
const functionsEmulator_1 = require("../emulator/functionsEmulator");
const emulatorServer_1 = require("../emulator/emulatorServer");
const functionsEmulatorUtils_1 = require("../emulator/functionsEmulatorUtils");
const getProjectId = require("../getProjectId");
class FunctionsServer {
    constructor() {
        this.emulatorServer = undefined;
    }
    assertServer() {
        if (!this.emulatorServer) {
            throw new Error("Must call start() before calling any other operation!");
        }
    }
    async start(options, partialArgs) {
        const projectId = getProjectId(options, false);
        const functionsDir = path.join(options.config.projectDir, options.config.get("functions.source"));
        const nodeMajorVersion = functionsEmulatorUtils_1.parseRuntimeVersion(options.config.get("functions.runtime"));
        const args = Object.assign({ projectId,
            functionsDir,
            nodeMajorVersion }, partialArgs);
        if (options.host) {
            args.host = options.host;
        }
        if (options.port) {
            const hostingRunning = options.targets && options.targets.indexOf("hosting") >= 0;
            if (hostingRunning) {
                args.port = options.port + 1;
            }
            else {
                args.port = options.port;
            }
        }
        this.emulatorServer = new emulatorServer_1.EmulatorServer(new functionsEmulator_1.FunctionsEmulator(args));
        await this.emulatorServer.start();
    }
    async connect() {
        this.assertServer();
        await this.emulatorServer.connect();
    }
    async stop() {
        this.assertServer();
        await this.emulatorServer.stop();
    }
    get() {
        this.assertServer();
        return this.emulatorServer.get();
    }
}
exports.FunctionsServer = FunctionsServer;
