// exegesis/lib/oas3/Oas3CompileContext.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ld = __importStar(require("lodash"));
const jsonPtr = __importStar(require("json-ptr"));
const json_schema_resolve_ref_1 = require("../utils/json-schema-resolve-ref");
/**
 * This has common stuff that we want to pass all the way down through the OAS
 * heirarchy.  This also keeps track of the `path` that a given object was
 * generated from.
 */
class Oas3CompileContext {
    constructor(a, path, options) {
        if (a instanceof Oas3CompileContext) {
            // TODO: Could make this WAY more efficient with Object.create().
            const parent = a;
            this.path = parent.path.concat(path);
            this.openApiDoc = parent.openApiDoc;
            this.options = parent.options;
        }
        else if (options) {
            this.path = path.slice();
            this.openApiDoc = a;
            this.options = options;
        }
        else {
            throw new Error('Invalid parameters to Oas3CompileContext constructor');
        }
        this.jsonPointer = jsonPtr.encodePointer(this.path);
    }
    childContext(relativePath) {
        if (ld.isArray(relativePath)) {
            return new Oas3CompileContext(this, relativePath);
        }
        else {
            return new Oas3CompileContext(this, [relativePath]);
        }
    }
    resolveRef(ref) {
        return json_schema_resolve_ref_1.resolveRef(this.openApiDoc, ref);
    }
}
exports.default = Oas3CompileContext;
